subsys/bluetooth/services/cgms/cgms.c

```diff
@@ -0,0 +1,455 @@
+			void *buf, uint16_t len, uint16_t offset)
+{
+	struct net_buf_simple *feature_buf = NET_BUF_SIMPLE(CGMS_FEAT_LENGTH);
+
+	net_buf_simple_init(feature_buf, 0);
+
+	net_buf_simple_add_le24(feature_buf, cgms_inst.feature.feature);
+	net_buf_simple_add_u8(feature_buf,
+		(cgms_inst.feature.type & 0xF) |
+		((cgms_inst.feature.sample_location & 0xF) << 4));
+	net_buf_simple_add_le16(feature_buf, 0xFFFF);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, feature_buf->data, feature_buf->len);
+}
+
+static ssize_t read_session_start_time(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			void *buf, uint16_t len, uint16_t offset)
+{
+	struct net_buf_simple *sst_buf = NET_BUF_SIMPLE(CGMS_SST_LENGTH);
+
+	net_buf_simple_init(sst_buf, 0);
+
+	net_buf_simple_add_le16(sst_buf, cgms_inst.sst.year);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.month);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.day);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.hour);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.minute);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.second);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.timezone);
+	net_buf_simple_add_u8(sst_buf, cgms_inst.sst.dst_offset);
+
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, sst_buf->data, sst_buf->len);
+}
+
+static bool leap_year_check(uint16_t year)
+{
+	if (year % 400 == 0) {
+		return true;
+	} else if (year % 100 == 0) {
+		return false;
+	} else if (year % 4 == 0) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+static int cal_sst_offset(struct net_buf_simple *sst_buffer)
+{
+	struct cgms_session_start_time sst;
+	uint32_t time_offset;
+	uint32_t day_to_change;
+	uint32_t second_to_change;
+	uint32_t residual_day;
+	uint32_t residual_second;
+	int day_of_year[13] = {0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365};
+	int day_of_leap_year[13] = {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366};
+
+	if (!sst_buffer || sst_buffer->size < 9 || sst_buffer->len < 9) {
+		return -EINVAL;
+	}
+
+	sst.year = net_buf_simple_pull_le16(sst_buffer);
+	sst.month = net_buf_simple_pull_u8(sst_buffer);
+	sst.day = net_buf_simple_pull_u8(sst_buffer);
+	sst.hour = net_buf_simple_pull_u8(sst_buffer);
+	sst.minute = net_buf_simple_pull_u8(sst_buffer);
+	sst.second = net_buf_simple_pull_u8(sst_buffer);
+	sst.timezone = net_buf_simple_pull_u8(sst_buffer);
+	sst.dst_offset = net_buf_simple_pull_u8(sst_buffer);
+
+	/* Get elapsed time in second */
+	time_offset = (k_uptime_get_32() - cgms_inst.local_start_time) / MSEC_PER_SEC;
+	day_to_change = time_offset / SEC_PER_DAY;
+	second_to_change = time_offset % SEC_PER_DAY;
+	residual_second = sst.hour * 3600 + sst.minute * 60 + sst.second;
+	if (residual_second < second_to_change) {
+		residual_second += SEC_PER_DAY;
+		day_to_change++;
+	}
+	/* hh:mm:ss */
+	sst.hour = (residual_second - second_to_change) / 3600;
+	sst.minute = (residual_second - second_to_change - sst.hour * 60) / 60;
```

Comment:
Shouldn't you use minute here?

```suggestion
	sst.minute = (residual_second - second_to_change - sst.minute * 60) / 60;
```

Reply:
You are right, as the unit here is second. But as we are calculating minute, it should be sst.hour * 3600.